torch/csrc/jit/codegen/cuda/expr_simplifier.cpp

```diff
@@ -0,0 +1,430 @@
+#include <torch/csrc/jit/codegen/cuda/expr_simplifier.h>
+
+#include <torch/csrc/jit/codegen/cuda/ir_all_nodes.h>
+#include <torch/csrc/jit/codegen/cuda/ir_builder.h>
+#include <torch/csrc/jit/codegen/cuda/ir_cloner.h>
+
+#include <unordered_set>
+#include <vector>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+
+namespace {
+
+bool hasSimilarType(DataType t1, DataType t2) {
+  if (t1 == t2) {
+    return true;
+  }
+  if (isIntegralType(t1) && isIntegralType(t2)) {
+    return true;
+  }
+  if (isFloatingPointType(t1) && isFloatingPointType(t2)) {
+    return true;
+  }
+  if (isComplexType(t1) && isComplexType(t2)) {
+    return true;
+  }
+  return false;
+}
+
+bool hasValue(Val* value) {
+  return value->getInt() != c10::nullopt ||
+      value->getDouble() != c10::nullopt || value->getBool() != c10::nullopt;
+}
+
+bool isMagicZeroProtectedVariable(Val* value) {
```

Comment:
Do we want to change magic zero to be something like a set carried in GpuLower to be applied in `codegen.cpp` instead of being an explicit expression? Could also make sense for magic zero protection being a unary op itself.

Reply:
I think what you suggested makes sense and would make it cleaner. But I am not sure if or when I will go to this direction, because I want to double check that, if with all my current index hoisting and expr simplification, nvfuser zero is still needed. If we luckily find that nvfuser zero is no longer needed, then we could just abandon it totally.